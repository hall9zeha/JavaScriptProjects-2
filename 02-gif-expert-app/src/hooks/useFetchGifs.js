import { useEffect, useState } from "react"
import { fetchGifs } from "../helpers/GetGifs";



export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    data:[],
    loading:true
  });

  //Los efectos o pueden ser asíncronos pero perfectamente se puede usar eventos asíncronos dentro de ellos
  useEffect(()=>{
    fetchGifs(category).
        then(imgs=>{
            setState({
                data:imgs,
                loading:false
            })
        });
    },[category]); //Esto hará que la función sea llamada solo la primera vez que se renderice el componente 
                   //y solo cuando cambie la categoría 


  return state;
}
