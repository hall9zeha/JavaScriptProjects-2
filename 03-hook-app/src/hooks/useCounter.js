import { useState } from "react";


export const useCounter = (initState = 10) => {
  const [state, setState] = useState(initState);


  const increment = (factor=1)=>{
    setState(state +factor);
  }
  const decrement = (factor=1)=>{
    setState(state -factor);
  }
  return {
    state,
    increment,
    decrement
  }
}
