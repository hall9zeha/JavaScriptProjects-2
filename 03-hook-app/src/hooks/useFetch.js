import { useEffect, useRef, useState } from "react"


export const useFetch = (url) => {

  const isMounted = useRef(true)//Para tomar como referencia al componente cuando esta montado o desmontado


  const[state, setState] = useState({data:null, loading:true, error:null});

  useEffect(()=>{
    return ()=>{
      isMounted.current = false;//Cuando el componente esté desmontado retornará  ese estado
    }
  },[])

  useEffect(()=>{
    setState({data:null,loading:true,error:null});
    fetch(url)
    .then(resp=>resp.json())
    .then(data=>{
      if(isMounted.current){//Si el componente todavía está montado retornar la data 
        setState({
            loading:false,
            error:null,
            data
        })
      }
    })


  },[url])
  return state;
}
