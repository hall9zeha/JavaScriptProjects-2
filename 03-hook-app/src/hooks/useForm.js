import { useState } from "react"


export const useForm = (initValues = {}) => {
    const[values, setValues] = useState(initValues);

    const resetForm = ()=>{
        setValues(initValues);
    }
    const handleOnChange =({target})=>{
        setValues({
            ...values,
            [target.name]:target.value
        })
    }

    return [values, handleOnChange, resetForm]
}
