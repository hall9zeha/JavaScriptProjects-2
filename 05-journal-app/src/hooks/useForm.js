import React, { useState } from 'react'

export const useForm = (initialstate={}) => {
    const [values, setValues] = useState(initialstate);

    const reset = () =>{
        setValues(initialstate)
    }

    const handleInputChange = ({target}) =>{
        setValues({
            ...values,
            [target.name]:target.value
        })
    }
    return [values,handleInputChange,reset];    
}


