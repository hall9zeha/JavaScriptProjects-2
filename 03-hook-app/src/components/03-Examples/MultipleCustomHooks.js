
import React from 'react'
import { useCounter } from '../../hooks/useCounter'
import { useFetch } from '../../hooks/useFetch'
import '../02-useEffect/effects.css'
export const MultipleCustomHooks = () => {

  const {state, increment, decrement} = useCounter(1)//Para incrementar la cantidad de quotes que queremos obtener
  const {loading, data} =  useFetch(`https://api.breakingbadquotes.xyz/v1/quotes/${state}`); 
  const listQuotes = !!data && data //Si la data es diferente de nulo asignamos la respuesta a la constante listQuotes
   
  return (
    <div>
      <h3>BreakingBad quotes</h3>
      <hr/>
      <div  data-testid='multiple-custom-hooks'>
      {
        loading ? (
          <div className='alert alert-info text-center'>
            Loading...
          </div>
        ) : 
        (
          listQuotes.map(quote =>(
                  <blockquote key={Math.random()} className='blockquote float-end'>
                      <p  className='mb-0'>{quote.quote}</p>
                      <footer className='blockquote-footer'>{quote.author}</footer>
                  </blockquote>
          ))
         
        )
      }
      </div>
      {/* En este ejemplo, se ha añadido un div vacío con un estilo clear: both entre la lista de elementos y los botones. 
      Esto garantizará que los botones siempre se coloquen después del último elemento creado por el mapeo
      y no se superpongan a los elementos anteriores. */}
      <div style={{ clear: 'both' }}></div>
      <div className='mt-3'>
      <button onClick={()=>{increment(1)}} className='btn btn-primary'>Traer más</button>
      <button onClick={()=>{decrement(1)}} className='btn btn-primary ms-2'>Traer menos</button>
      </div>
     </div>
  )
}
