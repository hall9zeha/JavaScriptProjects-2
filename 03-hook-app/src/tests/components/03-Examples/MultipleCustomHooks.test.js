import renderHook, {act, waitFor,render, screen} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { MultipleCustomHooks } from '../../../components/03-Examples/MultipleCustomHooks';
import { useCounter } from '../../../hooks/useCounter';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('Tests in MultipleCustomHooks', () => { 

    beforeEach(()=>{
        useCounter.mockReturnValue({
            state: 10,
            increment: jest.fn(),
            decrement: jest.fn(),
            reset: jest.fn(),
        })
    })
    
    
    test('should show successful', () => {
      
//Usamos mock para devolver data falsa, ya que no nos interesa la data real, solo 
//el funcionamiento de los hooks
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        })
        const wrapper = renderer.create(<MultipleCustomHooks/>).toJSON()
        expect(wrapper).toMatchSnapshot();

     })
    
    test('should show information',()=>{

        useFetch.mockReturnValue({
            data:[{
                author:"Barry",
                quote:"Hello World"
            }],
            loading:false,
            error:null
        })
        const {container}= render(<MultipleCustomHooks/>)
        expect(container.querySelector('.alert')).toBeNull()
        expect(screen.queryByRole('alert',{name:'Loading...'})).toBeNull()
        expect(container.querySelector('.mb-0').textContent).toBe('Hello World')
        expect(container.querySelector('footer').textContent).toBe('Barry')
        console.log(container.outerHTML)
    })

 })