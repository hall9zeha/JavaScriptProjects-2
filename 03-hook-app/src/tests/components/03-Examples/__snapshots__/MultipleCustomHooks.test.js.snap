// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in MultipleCustomHooks should show successful 1`] = `
<div>
  <h3>
    BreakingBad quotes
  </h3>
  <hr />
  <div
    data-testid="multiple-custom-hooks"
  >
    <div
      className="alert alert-info text-center"
    >
      Loading...
    </div>
  </div>
  <div
    style={
      Object {
        "clear": "both",
      }
    }
  />
  <div
    className="mt-3"
  >
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Traer m√°s
    </button>
    <button
      className="btn btn-primary ms-2"
      onClick={[Function]}
    >
      Traer menos
    </button>
  </div>
</div>
`;
